In [1]:
!pip install pandas numpy scikit-learn

In [2]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import mean_squared_error

In [3]:
df = pd.read_csv("./AmesHousing.csv")
df.head()

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,...,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
0,1,526301100,20,RL,141.0,31770,Pave,NaN,IR1,Lvl,...,0,NaN,NaN,NaN,0,5,2010,WD,Normal,215000
1,2,526350040,20,RH,80.0,11622,Pave,NaN,Reg,Lvl,...,0,NaN,MnPrv,NaN,0,6,2010,WD,Normal,105000
2,3,526351010,20,RL,81.0,14267,Pave,NaN,IR1,Lvl,...,0,NaN,NaN,Gar2,12500,6,2010,WD,Normal,172000
3,4,526353030,20,RL,93.0,11160,Pave,NaN,Reg,Lvl,...,0,NaN,NaN,NaN,0,4,2010,WD,Normal,244000
4,5,527105010,60,RL,74.0,13830,Pave,NaN,IR1,Lvl,...,0,NaN,MnPrv,NaN,0,3,2010,WD,Normal,189900


Variables con datos NaN ordenados de mayor a menor

In [4]:
df.isna().sum()[df.isnull().sum() > 0].sort_values(ascending=False)

,0
Pool QC,2917
Misc Feature,2824
Alley,2732
Fence,2358
Mas Vnr Type,1775
Fireplace Qu,1422
Lot Frontage,490
Garage Qual,159
Garage Cond,159
Garage Yr Blt,159


Eliminar colmnas con datos nulos

In [5]:
def eliminar_columnas_con_nulos(df):
    cols_con_nulos = df.columns[df.isnull().any()]
    df_limpio = df.drop(columns=cols_con_nulos)
    return df_limpio

In [6]:
df_limpio = eliminar_columnas_con_nulos(df)

Dividir el dataset en train y test

In [7]:
X = df_limpio.drop(columns=['SalePrice'])
y = df_limpio['SalePrice']

In [8]:
X = pd.get_dummies(X)

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Cargar modelo, entrenar y predecir

In [10]:
modelo = LinearRegression()
modelo.fit(X_train, y_train)
y_pred = modelo.predict(X_test)

Evaluación y MSE

In [11]:
for real, pred in zip(y_test.iloc[:10], y_pred[:10]):
    print(f"Valor real: {real} | Predicción: {pred}")

Valor real: 161000 | Predicción: 174003.69444689807
Valor real: 116000 | Predicción: 106561.37281232327
Valor real: 196500 | Predicción: 212141.81356655248
Valor real: 123600 | Predicción: 132174.48149517924
Valor real: 126000 | Predicción: 123599.52068424504
Valor real: 174190 | Predicción: 189564.9750368176
Valor real: 200000 | Predicción: 146668.45696113165
Valor real: 148500 | Predicción: 141079.59449854214
Valor real: 88750 | Predicción: 112608.76287493855
Valor real: 409900 | Predicción: 351285.1094098985


In [12]:
mse = mean_squared_error(y_test, y_pred)
print("MSE:", mse)

MSE: 1020339958.2264279


Importar librería "entrena_modelo_3.py", modelo SVM y MinMaxScaler

In [ ]:
from entrena_modelo_api3 import modelo_ml
from sklearn.svm import SVR
from sklearn.preprocessing import MinMaxScaler

In [15]:
modelo_svm = SVR(kernel='linear')

ml = modelo_ml(X, y)

resultado = ml.entrenar_modelo(
    modelo=modelo_svm,
    seed_number=0,
    funcion_escalado=MinMaxScaler,
    estimador='mean_absolute_error',
    tamano=0.2,
    escalar='SI',
    tipo='REGRESION',
    estratifica='NO'
)

Entrenamiento estimación :  53934.674931160036
Entrenamiento score R2 : -0.015049671609631998
Prueba estimación :  55249.5982558237
Prueba score R2 : -0.03031701542107279


In [16]:
modelo_svm = SVR(kernel='rbf')

ml = modelo_ml(X, y)

resultado = ml.entrenar_modelo(
    modelo=modelo_svm,
    seed_number=0,
    funcion_escalado=MinMaxScaler,
    estimador='mean_absolute_error',
    tamano=0.2,
    escalar='SI',
    tipo='REGRESION',
    estratifica='NO'
)

Entrenamiento estimación :  55694.10186057917
Entrenamiento score R2 : -0.062280637823873164
Prueba estimación :  57141.225550720286
Prueba score R2 : -0.08002498505457734


In [17]:
modelo_svm = SVR(kernel='poly')

ml = modelo_ml(X, y)

resultado = ml.entrenar_modelo(
    modelo=modelo_svm,
    seed_number=0,
    funcion_escalado=MinMaxScaler,
    estimador='mean_absolute_error',
    tamano=0.2,
    escalar='SI',
    tipo='REGRESION',
    estratifica='NO'
)

Entrenamiento estimación :  55608.3258117308
Entrenamiento score R2 : -0.0597673006404722
Prueba estimación :  57048.58810723351
Prueba score R2 : -0.07738138850066067


En kernel lineal presenta en MAE más bajo: 53935